websockets/test_client_server.py

```diff
@@ -442,18 +442,41 @@ def test_invalid_status_error_during_client_connect(self):
         self.assertEqual(exception.code, 403)
         self.stop_server()
 
-    @unittest.mock.patch('websockets.server.read_request')
-    def test_connection_error_during_opening_handshake(self, _read_request):
-        _read_request.side_effect = ConnectionError
+    def do_error_during_opening_handshake(self, exception_cls):
+        written = []
 
-        self.start_server()
+        class WriteCollectingProtocol(WebSocketServerProtocol):
+
+            def client_connected(self, reader, writer):
+                original_write = writer.write
+
+                def collecting_write(data):
+                    original_write(data)
+                    written.append(data)
+                writer.write = collecting_write
+                super().client_connected(reader, writer)
+
+            @asyncio.coroutine
+            def handshake(self, *args, **kwargs):
+                raise exception_cls()
+
+        self.start_server(klass=WriteCollectingProtocol)
         # Exception appears to be platform-dependent: InvalidHandshake on
         # macOS, ConnectionResetError on Linux. This doesn't matter; this
         # test primarily aims at covering a code path on the server side.
```

Comment:
FWIW I modified the following lines like this:
```python
        with self.assertRaises(Exception) as exc:
            self.start_client()
        print(exc.exception)
```
and I'm getting these exceptions:
```
Status code not 101: 500
Malformed HTTP message
```
on macOS.